Fix/8751 unicode filename fix for windows

jNullj opened the pull request:
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )

Fixes #8751

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
https://user-images.githubusercontent.com/15849761/201156634-09ed45d1-db32-4802-9695-44061c5f3947.mp4



## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Compiled and run the application on windows.
Attempted to reproduce both #8751 and #7139 and tested both are solved.
It could be neat if i could run automated testing on github without playing with my own environment...
edit: i see you got some actions running on pull, forget about the line above.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)

jNullj committed:
Add unicode support for windows db filenames

FindFirstFile doesn't support unicode in file names.
Switch to FindFirstFileEx to allow unicode in db file name.

Fixes #8751

jNullj committed:
Add unicode support for windows db filenames

Use FindFirstFileW for explicit unicode support.
Use curret types to avoid casting issues.

Fixes keepassxreboot#8751

jNullj committed:
Fixed issue where filenames could change

Solve filename changes as shown in issue #7139.
If filename input is not the same capitalization file extention was added to file name when saved.
This change fixes it.

Fixes #8751

phoerious commented on src/main.cpp:
```diff
@@ -120,11 +120,11 @@ int main(int argc, char** argv)
+        WIN32_FIND_DATAW findFileData;
         HANDLE hFind;
-        hFind = FindFirstFile(fileInfo.absoluteFilePath().toUtf8(), &findFileData);
+        hFind = FindFirstFileW(fileInfo.absoluteFilePath().toStdWString().c_str(), &findFileData);
```
Why are you not using `String::toWCharArray()` here?

phoerious replied:
Actually, doesn't matter. toWCharArray() would also require you to allocate a wchar_t* first, so this is fine, I suppose.

droidmonkey replied:
Wouldn't `utf16()` do the trick? Internally QString is stored as UTF-16.

> QString stores the string using the UTF-16 encoding. Each of the 2 bytes of UTF-16 is represented using a QChar. One main reason to use UTF-16 as the internal representation is that it makes it fast to use them with native unicode APIs' on the Mac OS X and Windows (which expect UTF-16).

> Unicode methods in QString
A QChar represents a unicode code point. QString::unicode() returns the QChars of a QString. QString::utf16() returns ushort *. Notice that the function is not named toUtf16() because there is no conversion involved since the internal representation of QString is UTF-16.

jNullj replied:
I am not sure this is more elegant, im getting
`argument of type "const ushort *" is incompatible with parameter of type "LPCWSTR"`
i would still have to convert to wchar_t...

droidmonkey replied:
Cast, not convert. reinterpret_cast&lt;const wchar_t&gt; should do it

jNullj replied:
i could replace it with the following:
```c++
        const wchar_t* absolutePathWchar = (wchar_t*)fileInfo.absoluteFilePath().utf16();
        hFind = FindFirstFileW(absolutePathWchar, &findFileData);
```
not sure if this is better, tested this to work, couldnt do it without casting to wchar_t

droidmonkey replied:
Yah that's the way but use modern casting

jNullj replied:
Made changes at commit 43c0080a0a6cf5e41fcb6635bd4b330304b54c9c

jNullj committed:
Change path obtaining method

Change to method obtaining absolute file path from toStdWString to utf16.
This avoids internal conversation in QString as it is stored as utf16 ushort.

Fixes #8751
